akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala

```diff
@@ -0,0 +1,128 @@
+
+  def createWorkExecutor(): ActorRef[DoWork] = {
+    val supervised = Behaviors.supervise(workExecutorFactory()).onFailure[Exception](SupervisorStrategy.stop)
+    val ref = ctx.spawn(supervised, "work-executor")
+    ctx.watch(ref)
+    ref
+  }
+
+  def reportWorkFailedOnRestart(
+      workId: String): PartialFunction[(scaladsl.ActorContext[Worker.Command], Signal), Behavior[Worker.Command]] = {
+    case (_, Terminated(_)) =>
+      ctx.log.info("Work executor terminated. Reporting failure")
+      masterProxy ! Master.WorkFailed(ctx.self, workId)
+      // need to re-create the work executor
+      idle(createWorkExecutor())
+  }
+
+  def idle(workExecutor: ActorRef[DoWork] = createWorkExecutor()): Behavior[Worker.Command] =
+    Behaviors.setup[Worker.Command] { ctx =>
+      Behaviors.receiveMessagePartial[Worker.Command] {
+        case WorkIsReady =>
+          // this is the only state where we reply to WorkIsReady
+          masterProxy ! WorkerRequestsWork(workerId, ctx.self)
+          Behaviors.same
+
+        case SubmitWork(Work(workId, job: Int)) =>
+          ctx.log.info("Got work: {}", job)
+          workExecutor ! WorkExecutor.DoWork(job, ctx.self)
+          working(workId, workExecutor)
+
+        case Ack(_) =>
+          Behaviors.same
+
+      }
+    }
+
+  def working(workId: String, workExecutor: ActorRef[DoWork]): Behavior[Worker.Command] =
+    Behaviors.setup { ctx =>
+      Behaviors
+        .receiveMessagePartial[Worker.Command] {
+          case Worker.WorkComplete(result) =>
+            ctx.log.info("Work is complete. Result {}.", result)
+            masterProxy ! Master.WorkIsDone(workerId, workId, result, ctx.self)
+            ctx.setReceiveTimeout(workAckTimeout, WorkTimeout)
+            waitForWorkIsDoneAck(result, workId, workExecutor)
+
+          case _: SubmitWork =>
+            ctx.log.warn("Yikes. Master told me to do work, while I'm already working.")
+            Behaviors.unhandled
+
+        }
+        .receiveSignal(reportWorkFailedOnRestart(workId))
+    }
+
+  def waitForWorkIsDoneAck(result: String, workId: String, workExecutor: ActorRef[DoWork]): Behavior[Worker.Command] =
+    Behaviors.setup { ctx =>
+      Behaviors
+        .receiveMessage[Worker.Command] {
+          case WorkTimeout =>
+            ctx.log.info("No ack from master, resending work result")
+            masterProxy ! Master.WorkIsDone(workerId, workId, result, ctx.self)
+            Behaviors.same
+          case Ack(id) if id == workId =>
+            ctx.log.info("Work acked")
+            masterProxy ! WorkerRequestsWork(workerId, ctx.self)
+            ctx.cancelReceiveTimeout()
+            idle(workExecutor)
+        }
+        .receiveSignal(reportWorkFailedOnRestart(workId))
+
+    }
+
+}
+
+/**
+ * The worker is actually more of a middle manager, delegating the actual work
+ * to the WorkExecutor, supervising it and keeping itself available to interact with the work master.
+ */
+object Worker {
```

Comment:
Maybe we should just rename it to `MiddleManager` then? (I'm at least half serious ;) )

A bit weird with the companion below the class IMO